subsys/fs/ext2/ext2_bitmap.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2023 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include "ext2_bitmap.h"
+
+LOG_MODULE_DECLARE(ext2);
+
+/* NOTICE: Offsets in bitmap start with 0 */
+
+int ext2_bitmap_get(uint8_t *bm, uint32_t index, uint32_t size)
+{
+	uint32_t idx = index / 8;
+	uint32_t off = index % 8;
+
+	if (idx >= size) {
+		LOG_ERR("(get) Tried to access value outside of bitmap (%d)", index);
+		return -EINVAL;
+	}
+
+	uint8_t b = bm[idx];
+
+	return b & BIT(off);
+}
+
+int ext2_bitmap_set(uint8_t *bm, uint32_t index, uint32_t size)
+{
+	LOG_DBG("Setting %d bit in bitmap", index);
+
+	uint32_t idx = index / 8;
+	uint32_t off = index % 8;
+
+	if (idx >= size) {
+		LOG_ERR("Tried to set value outside of bitmap (%d)", index);
+		return -EINVAL;
+	}
```

Comment:
This is not OK. Not that the test is completely wrong, test is fine, but this is internal function you have complete control over and you have complete control over what you pass here.
This function can fail for a single reason: getting incorrect argument from caller, which is controlled by you.
There are two places where the fail can come from: `ext2_alloc_inode` and `ext2_alloc_block`.
In both case code looks similar, something like:
```
       (...)
        bitmap_slot = ext2_bitmap_find_free(BGROUP_BLOCK_BITMAP(&fs->bgroup), fs->block_size);
	if (bitmap_slot < 0) {
		LOG_WRN("Cannot find free block in group %d (rc: %d)", group, bitmap_slot);
		return bitmap_slot;
	}

	/* In bitmap blocks are counted from s_first_data_block hence we have to add this offset. */
	total = group * fs->sblock.s_blocks_per_group + bitmap_slot + fs->sblock.s_first_data_block;

	LOG_DBG("Found free block %d in group %d (total: %d)", bitmap_slot, group, total);

	rc = ext2_bitmap_set(BGROUP_BLOCK_BITMAP(&fs->bgroup), bitmap_slot, fs->block_size);
	if (rc < 0) {
		return rc;
	}
        (...)
```
where `ext2_bitmap_set` is passed result of `ext2_bitmap_find_free`, above named  `bitmap_slot`; so before the `ext2_bitmap_set` the `bitmap_slot` is used even though it may not be good as `ext2_bitmap_set` will report later.
The `ext2_bitmap_find_free` is able to report itself bad value (making it negative), yet it can still return bad value as a positive.
Questions: 
1) is it even possible to return bad index by the `ext2_bitmap_find_free`, as it operates on the same bitmaps as the `ext2_bitmap_set`?
2) Is this check really needed?
3) If 1) and 2) is true: shouldn't the function just return error code and LOG_ERRor be happening in the caller code to indicate where the problem has been really spotted?

Reply:
I have to disagree with you on this one. There are several reasons why we should want to leave these checks in the code, even though they might seem redundant at first glance.

1. I acknowledge the fact that right now we have total control over how this function (and others that you’ve pointed out) are used. That being said, this code will most likely change in the future due to various reasons, as all code does, or the code that uses this function will. Having these checks here in place let us easily convey what values of input arguments are valid for these functions.
2. IMO checking the value of input arguments (low trust) is a good practice. You don’t delegate the responsibility of checking input arguments to some other place in the code. Right now we’re using this function in two places in a similar context. This might not always be the case. If we were to use this function in more contexts, we would basically duplicate the code responsible for sanitising the values of input arguments.
3. By having checks in the same place as the code that assumes that the conditions imposed by these checks are met, we don’t have to analyse the whole ext2 fs subsystem holistically. You have a discrete module (function), which you can easily analyse.
4. We want to be extra careful here. We don’t want to accidentally corrupt a filesystem.
5. As a bonus, this  can help in testing a function in isolation.